system/lib/wasmfs/syscalls.cpp

```diff
@@ -536,6 +540,81 @@ __wasi_errno_t __wasi_fd_fdstat_get(__wasi_fd_t fd, __wasi_fdstat_t* stat) {
   return __WASI_ERRNO_SUCCESS;
 }
 
+// This enum specifies whether rmdir or unlink is being performed.
+enum class UnlinkMode { Rmdir, Unlink };
+
+static long doUnlink(char* path, UnlinkMode unlinkMode) {
+  auto pathParts = splitPath(path);
+
+  // TODO: Ensure that . and .. are invalid when path parsing is updated.
+  // TODO: Change to check root directory pointer instead of path.
+  // This can be done when path parsing is refactored.
+  // Current state just matches JS file system behaviour.
+  if (pathParts.size() == 1 && pathParts[0] == "/") {
+    return -EBUSY;
+  }
+
+  if (pathParts.empty()) {
+    return -EINVAL;
+  }
+
+  auto base = pathParts.back();
+
+  long err;
+  auto parentDir = getDir(pathParts.begin(), pathParts.end() - 1, err);
+
+  // Parent node doesn't exist.
+  if (!parentDir) {
+    return err;
+  }
+
+  // Hold the locked directory to prevent the state from being changed during
+  // the operation.
+  auto lockedParentDir = parentDir->locked();
+
+  auto curr = lockedParentDir.getEntry(base);
+
+  if (!curr) {
+    return -ENOENT;
+  }
+
+  // rmdir checks if the target is a directory and if the directory is empty.
+  auto targetDir = curr->dynCast<Directory>();
+  if (unlinkMode == UnlinkMode::Rmdir) {
+
+    if (!targetDir) {
+      return -ENOTDIR;
+    }
+
+    // A directory can only be removed if it has zero entries.
+    if (targetDir->locked().getNumEntries() > 0) {
+      return -ENOTEMPTY;
+    }
+  } else {
+    // unlink cannot remove a directory.
+    if (targetDir) {
+      return -EISDIR;
+    }
+  }
+
+  // Cannot unlink/rmdir if the parent dir doesn't have write permissions.
+  if (!(lockedParentDir.mode() & WASMFS_PERM_WRITE)) {
+    return -EACCES;
+  }
+
+  // Erase targetDir from lockedParentDir and also set targetDir's parent
+  // pointer to nullptr.
```

Comment:
```suggestion
  // Input is valid, perform the unlink.
```
1. The two mentions of `targetDir` were inaccurate/stale.
2. There is no need to mention setting the parent pointer here. That is a detail of what the unlink method does. Imagine that in the future we don't have parent pointers, or some other refactoring, then `unlinkEntry()` would do something new and correct for the new model, but this comment would become stale and incorrect.